flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,183 @@
+		if (!schedulingVertex.isPresent()) {
+			throw new IllegalStateException("can not find scheduling vertex for " + executionVertexId);
+		}
+
+		if (ExecutionState.RUNNING.equals(executionState)) {
+			List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+			for (SchedulingResultPartition srp : schedulingVertex.get().getProducedResultPartitions()) {
+				if (srp.getPartitionType().isPipelined()) {
+					for (SchedulingVertex sv : srp.getConsumers()) {
+						executionVertexDeploymentOptions.add(
+							new ExecutionVertexDeploymentOption(sv.getId(), deploymentOption));
+					}
+				}
+			}
+			schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+		} else if (ExecutionState.FINISHED.equals(executionState)) {
+			List<SchedulingResultPartition> finishedPartitions = null;
+			for (SchedulingResultPartition srp : schedulingVertex.get().getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					Optional<SchedulingIntermediateDataSet> intermediateDataSet = schedulingTopology.getIntermediateDataSet(srp.getResultId());
+					if (!intermediateDataSet.isPresent()) {
+						throw new IllegalStateException("can not find scheduling intermediate dataset for " + srp.getResultId());
+					}
+					final int refCnt = intermediateDataSet.get().decrementNumberOfRunningProducersAndGetRemaining();
+					if (refCnt < 0) {
+						throw new IllegalStateException("Decremented number of unfinished producers below 0. "
+							+ "This is most likely a bug in the execution state/intermediate result "
+							+ "partition management.");
+					} else if (refCnt == 0) {
+						if (finishedPartitions == null) {
+							finishedPartitions = new LinkedList<>();
+						}
+						finishedPartitions.add(srp);
+					}
+				}
+			}
+
+			if (finishedPartitions != null) {
+				List<SchedulingVertex> toScheduleConsumers = new LinkedList<>();
+				for (SchedulingResultPartition finishedPartition : finishedPartitions) {
+					toScheduleConsumers.addAll(finishedPartition.getConsumers());
+				}
+				scheduleBlockingConsumers(toScheduleConsumers);
+			}
+		}
+	}
+
+	@Override
+	public void onPartitionConsumable(ExecutionVertexID executionVertexId, ResultPartitionID resultPartitionId) {
+		// No need to react to this notification.
+	}
+
+	private void scheduleBlockingConsumers(List<SchedulingVertex> vertices) {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new LinkedList<>();
+		for (SchedulingVertex schedulingVertex : vertices) {
+			final ExecutionState state = schedulingVertex.getState();
+
+			if (CREATED.equals(state)) {
+				final InputDependencyConstraint inputDependency = jobGraph.findVertexByID(schedulingVertex
+					.getJobVertexId()).getInputDependencyConstraint();
+
+				boolean shouldSchedule = true;
+				if (!InputDependencyConstraint.ANY.equals(inputDependency)) {
+					for (SchedulingResultPartition srp : schedulingVertex.getConsumedResultPartitions()) {
```

Comment:
This check loop can optimized during collecting the SchedulingVertex to schedule in the loop above.

Reply:
These two loops can not be merged since the first loop check the produced partitions while the second check the consumed partitions.